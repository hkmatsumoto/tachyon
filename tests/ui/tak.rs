fn tak(x: i32, y: i32, z: i32) -> i32 {
    if x <= y {
        z
    } else {
        tak(
            tak(x - 1, y, z),
            tak(y - 1, z, x),
            tak(z - 1, x, y),
        )
    }
}
