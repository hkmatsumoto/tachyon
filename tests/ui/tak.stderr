; ModuleID = 'top'
source_filename = "top"

define i32 @tak(i32 %x, i32 %y, i32 %z) {
entry:
  %_0 = alloca i32, align 4
  %_4 = alloca i1, align 1
  %_5 = alloca i32, align 4
  %_6 = alloca i32, align 4
  %_7 = alloca i32, align 4
  %_8 = alloca i32, align 4
  %_9 = alloca i32, align 4
  %_10 = alloca i32, align 4
  %_11 = alloca i32, align 4
  %_12 = alloca i32, align 4
  %_13 = alloca i32, align 4
  %_14 = alloca i32, align 4
  %_15 = alloca i32, align 4
  %_16 = alloca i32, align 4
  %_17 = alloca i32, align 4
  %_18 = alloca i32, align 4
  %_19 = alloca i32, align 4
  %_20 = alloca i32, align 4
  %_21 = alloca i32, align 4
  br label %bb0

bb0:                                              ; preds = %entry
  store i32 %x, i32* %_5, align 4
  store i32 %y, i32* %_6, align 4
  %0 = load i32, i32* %_5, align 4
  %1 = load i32, i32* %_6, align 4
  %2 = icmp sle i32 %0, %1
  store i1 %2, i1* %_4, align 1
  %3 = load i1, i1* %_4, align 1
  switch i1 %3, label %bb1 [
    i1 false, label %bb2
  ]

bb1:                                              ; preds = %bb0
  store i32 %z, i32* %_0, align 4
  br label %bb6

bb2:                                              ; preds = %bb0
  store i32 %x, i32* %_9, align 4
  %4 = load i32, i32* %_9, align 4
  %5 = sub i32 %4, 1
  store i32 %5, i32* %_8, align 4
  store i32 %y, i32* %_10, align 4
  store i32 %z, i32* %_11, align 4
  %6 = load i32, i32* %_8, align 4
  %7 = load i32, i32* %_10, align 4
  %8 = load i32, i32* %_11, align 4
  %9 = call i32 @tak(i32 %6, i32 %7, i32 %8)
  store i32 %9, i32* %_7, align 4
  br label %bb3

bb3:                                              ; preds = %bb2
  store i32 %y, i32* %_14, align 4
  %10 = load i32, i32* %_14, align 4
  %11 = sub i32 %10, 1
  store i32 %11, i32* %_13, align 4
  store i32 %z, i32* %_15, align 4
  store i32 %x, i32* %_16, align 4
  %12 = load i32, i32* %_13, align 4
  %13 = load i32, i32* %_15, align 4
  %14 = load i32, i32* %_16, align 4
  %15 = call i32 @tak(i32 %12, i32 %13, i32 %14)
  store i32 %15, i32* %_12, align 4
  br label %bb4

bb4:                                              ; preds = %bb3
  store i32 %z, i32* %_19, align 4
  %16 = load i32, i32* %_19, align 4
  %17 = sub i32 %16, 1
  store i32 %17, i32* %_18, align 4
  store i32 %x, i32* %_20, align 4
  store i32 %y, i32* %_21, align 4
  %18 = load i32, i32* %_18, align 4
  %19 = load i32, i32* %_20, align 4
  %20 = load i32, i32* %_21, align 4
  %21 = call i32 @tak(i32 %18, i32 %19, i32 %20)
  store i32 %21, i32* %_17, align 4
  br label %bb5

bb5:                                              ; preds = %bb4
  %22 = load i32, i32* %_7, align 4
  %23 = load i32, i32* %_12, align 4
  %24 = load i32, i32* %_17, align 4
  %25 = call i32 @tak(i32 %22, i32 %23, i32 %24)
  store i32 %25, i32* %_0, align 4
  br label %bb6

bb6:                                              ; preds = %bb5, %bb1
  %26 = load i32, i32* %_0, align 4
  ret i32 %26
}
